src/uproot/interpretation/objects.py

```diff
@@ -437,29 +437,54 @@ def final_array(
         )
         trimmed = []
         start = entry_offsets[0]
+        has_any_awkward_types = any(
+            uproot._util.from_module(x, "awkward") for x in basket_arrays.values()
+        )
```

Comment:
In the proposed change:
  if we were to have library = ak and only NumPy types, then it will go into the if clause and turn all the NumPy types into Awkward types by relying on a form that doesn’t exist; instead of simply appending all the NumPy types into `trimmed` which will then get concatenated and transformed into Awkward through line:  https://github.com/scikit-hep/uproot5/blob/1cee9606748482c8a06828646144bd44caac4b8f/src/uproot/interpretation/objects.py#L487

Reply:
I understand the argument now: if this array will _eventually_ convert from NumPy `dtype=object` into Awkward but none of the basket arrays are Awkward, we can let the concatenation happen at the level of NumPy arrays (very inexpensive, since they're arrays of Python _references_) and later code is already in place to do the conversion to Awkward all at once. That's the normal case, so this is an important optimization.